ui automation: global settings

yonasberhe23 opened the pull request:
<!-- This template is for Devs to give QA details before moving the issue To-Test -->
### Summary
This pr covers the functionality within Global Settings (Branding, Home-links, etc.)
View Test Plan [here](https://github.com/rancher/qa-tasks/issues/740)

### Additional Comments
I needed to split the Settings test into 2 spec files because Cypress browser crashes when there are a lot of tests in the spec file (there are 21 Settings tests total) - it seems like if there are more than 10 tests to 1 spec file then the browser crashes. This solution seems to have worked.
If we would like to have all 21 tests live in 1 spec file we would need to upgrade Cypress to at least v12.4.0 to use experimentalMemoryManagement config to help manage memory. See related issue https://github.com/cypress-io/cypress/issues/25949

~~There are a number of open bugs that will cause tests to fail:~~
- https://github.com/rancher/dashboard/issues/9551
- https://github.com/rancher/dashboard/issues/9444
- https://github.com/rancher/dashboard/issues/9690

richard-cox commented:
Moving to 2.8.next1 as per dependent issues

bmdepesa commented:
@richard-cox can you clarify why this test can't be merged now? 

From a QA perspective, it's better to have a test that's correctly failing due to issues than to wait on merging until we know it will pass.

richard-cox commented:
> @richard-cox can you clarify why this test can't be merged now?
> 
> From a QA perspective, it's better to have a test that's correctly failing due to issues than to wait on merging until we know it will pass.

@bmdepesa These will be part of PR CI, so any failures will block merging of all PRs.

Haven't dived in to them, but if there are some that succeed at the moment, the one's that don't can be disabled and enabled as part of the PRs that fix the underlying issues?

bmdepesa commented:
Thanks @richard-cox, that makes sense. 

We are investigating some methods to tag known failures so we can skip those in the GH Actions, but also routinely run the known failures in another workflow so we can better stay on top of which tests are fixed. We can let this sit for now.

richard-cox commented:
@yonasberhe23 I think master now has fixes for the three issues listed in the description. Is this PR now ready to merge (minus conflict)?

yonasberhe23 commented:
> @yonasberhe23 I think master now has fixes for the three issues listed in the description. Is this PR now ready to merge (minus conflict)?

@richard-cox I will need to add feature tag to these and fix the conflict but after that should be ready. will work on that today

izaac commented:
According to the description this is still on hold for a bug fix? Are there any updates on when the fix is getting merged?

yonasberhe23 commented:
@richard-cox @izaac The test related to that bug is currently being skipped so it shouldn't be on hold for it. My plan is to remove .skip from the test once its been resolved https://github.com/rancher/dashboard/blob/c53437606f60f7af874109bceb49f48286562ec5/cypress/e2e/tests/pages/global-settings/settings-p2.spec.ts#L20-L23

izaac commented:
@richard-cox that sounds like a good idea. In regards of the `To-Test` label I will follow up to understand it better.

richard-cox commented on cypress/e2e/po/pages/global-settings/banners.po.ts:
```diff
@@ -0,0 +1,119 @@
+
+  // input
+  headerInput(): LabeledInputPo {
+    return new LabeledInputPo(cy.getId('text-area-auto-grow').eq(0));
```
I think i'm derping, isn't `eq` equals, or does it select the first instance of an array of cypress elements?

yonasberhe23 replied:
its the later. it grabs a dom element at a specific index in an array of elements. in this case, we're targeting the first of three elements (with id `text-area-auto-grow`) found on the page

richard-cox reviewed (changes requested):
- can we use an svg instead of png for the custom-favicon? Once images merge to git they can be a pain
- same goes for rancher-dark and like grey, not 100% svgs are accepted, if not can the images be shrunk down to be super small?
- Worth a quick validation that all the new tests are correctly running in gh
- some of the skipped tests can now be enabled
- there's lots of goTo --> navTo changes requested. i think this is a newer pattern that we didn't have when this was originally created

yonasberhe23 commented on shell/pages/c/_cluster/settings/DefaultLinksEditor.vue:
```diff
@@ -81,6 +81,7 @@ export default {
             v-if="!isView"
             v-model="row.enabled"
             label-key="customLinks.settings.showLabel"
+            :data-testid="`checkbox-group-child-${i}`"
```
updated to `data-testid="custom-links__checkbox-${i}"`. how's that?

yonasberhe23 commented on shell/list/management.cattle.io.setting.vue:
```diff
@@ -92,6 +92,7 @@ export default {
       v-for="setting in settings"
       :key="setting.id"
       class="advanced-setting mb-20"
+      :data-testid="`advanced-setting-${setting.id}`"
```
updated to `data-testid="advanced-setting__option-${setting.id}"`. how's that?

yonasberhe23 commented on cypress/e2e/po/components/checkbox-input.po.ts:
```diff
@@ -15,8 +15,8 @@ export default class CheckboxInputPo extends ComponentPo {
    */
-  set(): Cypress.Chainable {
-    return this.input().click({ force: true });
+  set(optionIndex = 0): Cypress.Chainable {
```
good call! updated here https://github.com/rancher/dashboard/pull/9458/commits/4758cc5c218f587e82076c57abd32fe90b407461

izaac commented on cypress/e2e/po/edit/settings.po.ts:
```diff
@@ -0,0 +1,51 @@
+    cy.intercept('PUT', `/v1/management.cattle.io.settings/${ endpoint }`).as(interceptName);
+    this.saveButton().click();
+
+    return cy.wait(`@${ interceptName }`, { timeout: 15000 });
```
@yonasberhe23 I can help on understanding what's the reason of waiting this long in this particular Rancher area. I'm aware of some areas that usually take long and I can check if this one of them

yonasberhe23 replied:
Thanks, @izaac. I can't remember why I chose to use 15 sec timeout here but I've reduced it to 10 sec and tests are passing locally with that change. Let's see how they do in GA.

yonasberhe23 replied:
tests are passing with 10 sec timeout

yonasberhe23 commented on cypress/e2e/tests/pages/global-settings/peformance.spec.ts:
```diff
@@ -3,6 +3,9 @@ import { SettingsPagePo } from '@/cypress/e2e/po/pages/global-settings/settings.
 import HomePagePo from '@/cypress/e2e/po/pages/home.po';
```
I'm having trouble with this. The API request results in a 409 error in the `afterAll` step. do you know of a way to resolve this? see my changes here https://github.com/rancher/dashboard/pull/9458/commits/b8d916597034fccf20b251263b1ba406dc52305d

richard-cox replied:
Ahh ok, floor in my original request. 409s are conflict errors, which checks the revision of a resource is the same as previous when saving. Might need to remove that property from the update later on

yonasberhe23 replied:
okay, i fixed it by doing fetching the most recent resourceVersion and used that in the original request body to reset the settings to the initial step. fixed in https://github.com/rancher/dashboard/pull/9458/commits/5adbe40681759b1a5d835bfd5fe73b42c2014558

yonasberhe23 commented on cypress/e2e/tests/pages/global-settings/settings-p2.spec.ts:
```diff
@@ -0,0 +1,418 @@
+  });
+
+  it.skip('can update server-url', { tags: ['@globalSettings', '@adminUser'] }, () => {
+    // Note: this test fails sometimes due to https://github.com/rancher/dashboard/issues/9690
```
@richard-cox This is still an issue for me when running on my local machine on 2.8.0. I'm afraid if I remove the comment/`.skip` it will fail in CI

yonasberhe23 replied:
Please see my previous comment https://github.com/rancher/dashboard/pull/9458#issuecomment-1854424290

richard-cox replied:
The issue was fixed in 2.9.0, these changes will be going into master which currently tracks 2.9.0 so should now work

yonasberhe23 replied:
ah I see. re-enabled the test in https://github.com/rancher/dashboard/pull/9458/commits/5b48fbe571c025a4b3b6f2f12b56b871f4345c88

richard-cox commented on cypress/e2e/tests/pages/global-settings/banners.spec.ts:
```diff
@@ -0,0 +1,258 @@
+  });
+
+  it('can show and hide Header Banner', { tags: ['@globalSettings', '@adminUser'] }, () => {
+    bannersPage.goTo();
```
do these need to be goTo?

yonasberhe23 replied:
Nice catch, I missed this file. fixed in https://github.com/rancher/dashboard/pull/9458/commits/7112368f8cf9ac062e0b74643ed3fefd04b3b551

richard-cox commented on cypress/e2e/tests/pages/global-settings/branding.spec.ts:
```diff
@@ -94,22 +92,114 @@ describe('Branding', () => {
+    brandingPage.logoPreview('light').should('be.visible');
+
+    // Set dashboard theme to Dark and check top-level navigation header for updated logo in dark mode
+    prefPage.goTo();
```
could goTo's in this test be navTos?

yonasberhe23 replied:
updated

richard-cox commented on cypress/e2e/tests/pages/global-settings/settings-p2.spec.ts:
```diff
@@ -0,0 +1,419 @@
+    settingsPage.settingsValue('server-url').contains(settings['server-url'].new);
+
+    // Check Account and API Keys page
+    accountPage.goTo();
```
are the goTo's here needed?

yonasberhe23 replied:
Updated to `navTo`s. I was able to make it so that there is only one `goTo` left (besides the one in the `beforeAll` hook) for navigating to the cluster list page (`clusterList.goTo()`). That one needs to be there bc when updating it to `clusterList.navTo()` it causes issues with subsequent tests

richard-cox commented on cypress/e2e/po/pages/global-settings/branding.po.ts:
```diff
@@ -42,13 +55,26 @@ export class BrandingPagePo extends RootClusterPage {
     this.applyButton().click();
-    cy.wait(`@${ endpoint }`).its('response.statusCode').should('eq', 200);
+
+    return cy.wait(`@${ endpoint }`);
```
this drops the check for 200. would we be able to keep it, if not were there specific scenarios why it should be removed?

yonasberhe23 replied:
The `Favicon` test required property data checks (in the spec level) which couldn't be done while the `applyAndWait` fn had the status code check in the po which is why I removed it. I can add the status code checks on the test level where i'm calling `applyAndWait` and not checking status code. that work?

richard-cox replied:
Something like this should cover all bases. If ok, later on, we should use the same pattern in other places

```
  applyAndWait(endpoint: string, statusCode = 200): Cypress.Chainable {
    cy.intercept('PUT', endpoint).as(endpoint);
    this.applyButton().click();

    const wait = cy.wait(`@${ endpoint }`)

    return statusCode ? wait.its('response.statusCode').should('eq', wait) : wait;
  }
```

yonasberhe23 replied:
thanks for the suggestion! updated in https://github.com/rancher/dashboard/pull/9458/commits/c7361c8b9177a820284ed18c7ab5a3d953d955e9

richard-cox replied:
with the switch from the default statusCode of 200 to optional it now needs to be supplied

`applyAndWait(endpoint: string, statusCode = 200)` --> `applyAndWait(endpoint: string, statusCode?: number)`

`applyAndWait('xyz')` --> `applyAndWait('xyz', 200)`

i think we should get this pr in, so no need to update now, but worth considering in the future

richard-cox commented on cypress/e2e/po/pages/global-settings/home-links.po.ts:
```diff
@@ -0,0 +1,60 @@
+    cy.intercept('PUT', endpoint).as(endpoint);
+    this.applyButton().apply();
+
+    return cy.wait(`@${ endpoint }`);
```
This one could also check for a 200 response

yonasberhe23 replied:
added status code check here.

richard-cox commented on cypress/e2e/po/pages/global-settings/performance.po.ts:
```diff
@@ -27,13 +36,44 @@ export class PerformancePagePo extends RootClusterPage {
+    cy.intercept('PUT', 'ui-performance').as(`apply${ value }`);
+    this.applyButton().click();
+
+    return cy.wait(`@apply${ value }`);
```
this one too, for a 200

yonasberhe23 replied:
same as above comment https://github.com/rancher/dashboard/pull/9458#discussion_r1460150799

richard-cox reviewed (changes requested):
There's a few items left from https://github.com/rancher/dashboard/pull/9458#pullrequestreview-1600257507, and a few new ones. Getting close though!

yonasberhe23 commented:
> * can we use an svg instead of png for the custom-favicon? Once images merge to git they can be a pain
> * ~same goes for rancher-dark and like grey, not 100% svgs are accepted, if not can the images be shrunk down to be super small?~
> * Worth a quick validation that all the new tests are correctly running in gh
>   
>   * it doesn't look like @globalSettings are running?
> * ~some of the skipped tests can now be enabled~
> * ~there's lots of goTo --> navTo changes requested. i think this is a newer pattern that we didn't have when this was originally created~ (i've added comments to a few specific places)

@richard-cox 
- I've verified that all @globalSettings tests are running for both users; latest run http://139.59.134.103/run/7679e83a757144c8b68d931a5d390691
- updated the custom-favicon; now using svg instead of png

yonasberhe23 commented on cypress/e2e/tests/pages/global-settings/settings-p2.spec.ts:
```diff
@@ -0,0 +1,418 @@
+    const settingsEdit = settingsPage.editSettings('_', 'server-url');
+
+    settingsEdit.waitForPage();
+    cy.contains('Setting: server-url').should('be.visible');
```
updated in https://github.com/rancher/dashboard/pull/9458/commits/618ea7e0b8cd8ddf765bc973d777adc94f4fc267
